feat: rearranging standard fields in customize form

dj12djdjs opened the pull request:
continuation of: https://github.com/frappe/frappe/pull/19452

Customize Form now supports reordering all fields (including "standard" ones) 🎉 

### Implementation
A new property setter called `field_order` has been introduced. When `meta.sort_fields` is called following is done:
- if field order doesn't exist, existing algorithm applies

If field order exists:
- field order is "cleaned" for dead fields
- if it matches size of fields, new order is immediately updated (best case scenario)

In case new fields are present in doctype, but aren't in cleaned field order:
- if standard fields at the top of the DocType's fields table are absent in the field order, a list is created of such fields.
- if there are atleast some standard fields in field order, the above list is prepended to field order, else, the above list becomes field order (essentially field order gets invalidated)
- remaining fields are inserted using following logic:
  - `insert_after` for custom fields (same as before)
  - computed `insert_after` for standard fields (this is currently done based on the DocType's fields table)

### Additional implementation (if reviewer requests it)
- [ ] Delete invalid `field_order` (if no standard fields found) by enqueue-ing a DB call
- [ ] Update field order (when re-computed) by enqueue-ing a DB call
- [ ] Notify system managers if field order needs update. The "computed insert after" logic isn't 100% accurate so it can be beneficial if user reviews the order and manually fixes as needed (needs above update feature as well - to prevent duplicate notifications)

cc: @ankush @surajshetty3416 

### TODO (after initial review)
- [ ] tests for "computed" logic
- [x] support for form builder
- [ ] documentation

To be backported together with #20168

Commit:
feat: rearranging standing fields

Commit:
fix: fixed creation of property setter

Commit:
refactor: renamed setup_sortable

Commit:
fix: loading field_order property

Commit:
refactor: removed redundant db call

Aradhya-Tripathi committed:
Merge branch 'develop' into customize_form

Commit:
fix: field_order not found

Aradhya-Tripathi committed:
Merge branch 'frappe:develop' into customize_form

Commit:
test: Added tests for field order in customize form

Commit:
refactor: better naming

Commit:
refactor: simplified logic

Commit:
feat: Updating field order on custom field creation

Aradhya-Tripathi committed:
feat: Added support for custom fiels

Aradhya-Tripathi committed:
refactor: moving to meta

Aradhya-Tripathi committed:
Merge branch 'develop' into customize_form

Aradhya-Tripathi committed:
refactor: changed property type to json

Aradhya-Tripathi committed:
Merge branch 'customize_form' of github.com:Aradhya-Tripathi/frappe into customize_form

dj12djdjs committed:
Merge remote-tracking branch 'frappe/develop' into customize_form

dj12djdjs committed:
Merge remote-tracking branch 'frappe/develop' into customize_form

dj12djdjs committed:
fix: new standard field insert order.

dj12djdjs commented:
> system generated custom fields need to be treated like standard fields and property setters need to be created for those fields. since the v14 split, there are a lot of custom fields and if app developers make some changes, they have potential to override user customization.

@sagarvora Can you elaborate a bit more on this? I've tested by creating a system generated custom field in a custom app on the ToDo doctype. I then uninstalled the custom app, reset the database, I rearranged the fields, then installed the custom app. The system generated field is placed in the correct position based on: https://github.com/frappe/frappe/pull/19822/commits/6cdbe42f28d5944165dd100a9bb9172463951fda

Since this worked, I'm not sure what to test for.

pmjd-code commented:
I can offer to help with documentation if needed, would be great to see this implemented.

dj12djdjs commented:
> I can offer to help with documentation if needed, would be great to see this implemented.

@pmjd-code Thank you, I think we better wait until code changes are approved before writing documentation.

sagarvora commented:
> This seems to be already setup using a property setter tested by: [c212afb](https://github.com/frappe/frappe/commit/c212afbb9a84f17a541c0da8388c30974526c100)

Not really. [The subsequent code in `sort_fields` which creates `insert_after_map`](https://github.com/frappe/frappe/blob/6cdbe42f28d5944165dd100a9bb9172463951fda/frappe/model/meta.py#L492) will take precedence.

sagarvora commented:
> Can you elaborate a bit more on this? I've tested by creating a system generated custom field in a custom app on the ToDo doctype. I then uninstalled the custom app, reset the database, I rearranged the fields, then installed the custom app. The system generated field is placed in the correct position based on: [6cdbe42](https://github.com/frappe/frappe/commit/6cdbe42f28d5944165dd100a9bb9172463951fda)

Consider field [`gstin`](https://github.com/resilient-tech/india-compliance/blob/next/india_compliance/gst_india/constants/custom_fields.py#L19) in India Compliance.

1. User installed app.
2. User moved field to top of form for easier access. This caused `field_order` and `insert_after` to update.
3. User updated app, causing [fields to resync](https://github.com/resilient-tech/india-compliance/blob/510dab761a91a2f8a0c7332a826afe63ade242e0/india_compliance/patches.txt#L6).
4. Since `insert_after` gets the priority right now, user will need to update that value again in the Custom Field form.

This problem actually extends beyond order to other properties like `read_only` as well.

dj12djdjs committed:
Merge remote-tracking branch 'frappe/develop' into customize_form

dj12djdjs committed:
fix: don't modify insert_after of system generated custom fields.

# This is because system generated fields are to be treated as standard fields. If the user restores the form to default, this value will be used to reset the original position.

# The new position of form fields are stored in the field_order Property Setter.

dj12djdjs committed:
fix: treat system generated fields as standard fields when sorting.

dj12djdjs committed:
revert: check for is_system_generated

dj12djdjs committed:
Revert "fix: new standard field insert order."

This reverts commit 6cdbe42f28d5944165dd100a9bb9172463951fda.

dj12djdjs committed:
fix: prioritize field_order over insert_after.

# Use insert_after as fallback in event the field doesn't exist in field_order

dj12djdjs committed:
fix(test): delete existing custom field

dj12djdjs committed:
fix: order of standard fields without field_order property.

dj12djdjs committed:
Merge branch 'develop' into customize_form

dj12djdjs committed:
Revert "Revert "fix: new standard field insert order.""

This reverts commit c830f1ba2fe1e602b09c11a897869d9992097c1e.

dj12djdjs committed:
test: field order of newly migrated standard fields.

dj12djdjs committed:
fix(test): clear test_standard_field from previous test run.

dj12djdjs committed:
fix: sort with insert_after for system generated fields.

dj12djdjs committed:
fix(test): reset standard field creation before re-run and after successful test.

dj12djdjs committed:
fix: insert_after position should be + 1

dj12djdjs committed:
chore: remove debug statement

dj12djdjs committed:
test: system generated customized fields

dj12djdjs committed:
chore: remove print

pmjd-code commented:
Comment to stop it going stale.

sagarvora commented:
> system generated custom fields need to be treated like standard fields and property setters need to be created for those fields. since the v14 split, there are a lot of custom fields and if app developers make some changes, they have potential to override user customization.

this has been fixed separately here: https://github.com/frappe/frappe/pull/20168

rtdany10 commented:
@dj12djdjs @sagarvora will this be backported?

pmjd-code commented:
@rtdany10 this PR was to extend the recent Form Builder added to the develop branch, and as far as I'm aware the Form Builder won't be backported.
@dj12djdjs and @sagarvora will this PR now need #20168 to be merged before it can progress further?

sagarvora commented:
> will this PR now need #20168 to be merged before it can progress further?

I would definitely prefer that to avoid conflicts.

> will this be backported?

I think this piece independent of the form builder one should definitely be backported, but will discuss with the maintainers once.

sagarvora commented:
@dj12djdjs Thanks a lot for all the work you've done so far. I will be resuming on this now that https://github.com/frappe/frappe/pull/20168 has been merged. I am currently working on a separate branch, but if you're okay with it, I'd be happy to merge that branch into this one.

dj12djdjs commented:
@sagarvora No problem, I sent you permission for write access.

dj12djdjs committed:
Merge branch 'develop' into customize_from-sync

# Conflicts:
#	frappe/custom/doctype/customize_form/customize_form.js
#	frappe/custom/doctype/customize_form/test_customize_form.py

dj12djdjs committed:
chore: lint all

sagarvora committed:
Merge branch 'develop' into customize_form

sagarvora committed:
fix: show quick link to Table MultiSelect DocTypes

sagarvora committed:
refactor: change backend implementation of `CustomizeForm` and `Meta`

sagarvora committed:
test: simplify tests

sagarvora committed:
Merge branch 'develop' into customize_form

sagarvora commented:
### Additional implementation (if reviewer requests it)
- [ ] Delete invalid `field_order` (if no standard fields found) by enqueue-ing a DB call
- [ ] Update field order (when re-computed) by enqueue-ing a DB call
- [ ] Notify system managers if field order needs update. The "computed insert after" logic isn't 100% accurate so it can be beneficial if user reviews the order and manually fixes as needed (needs above update feature as well - to prevent duplicate notifications)

cc: @ankush @surajshetty3416

sagarvora committed:
fix: rename `idx` to `index` for clarity

sagarvora committed:
perf: define `existing_fields` conditionally

dj12djdjs committed:
Merge branch 'develop' into customize_form

sagarvora committed:
Merge branch 'develop' into customize_form

sagarvora committed:
Merge branch 'develop' into customize_form

pmjd-code commented:
The Form Builder is great and this feature would really help in customising apps. 
I know there is a lot going on and limited resources but does anyone know if this feature likely to make it soon or is it more likely months away?

sagarvora commented:
I will complete the remaining TODOs over the weekend and merge this.

If anyone has inputs on the additional implementation points above, they'd be greatly appreciated.

sagarvora commented:
I haven't been able to get this done. Some other projects that are more critical are taking longer. I don't have a new ETA as of now.

pmjd-code commented:
@sagarvora I understand. Thank you for the update and all of the hard work from you and @dj12djdjs on this.

If it is only the 4 files listed on thir PR needed to implement, I will probably try and add them in my own install so I can customise my installation of ERPnext more easily.

Thanks

ankush commented on frappe/custom/doctype/customize_form/customize_form.js:
```diff
@@ -87,23 +79,16 @@ frappe.ui.form.on("Customize Form", {
 	},
 
-	setup_sortable: function (frm) {
+	add_customize_child_table_button: function (frm) {
```
Note to self. This needs to be supported in form builder too. 

@shariquerik

sagarvora replied:
added it as a todo. I will do it.

ankush commented on frappe/model/meta.py:
```diff
@@ -452,14 +448,56 @@ def init_field_caches(self):
+		if field_order := getattr(self, "field_order", []):
+			field_order = [fieldname for fieldname in json.loads(field_order) if fieldname in self._fields]
+
+			# all fields match, best case scenario
+			if len(field_order) == len(self.fields):
+				self._update_fields_based_on_order(field_order)
```
This is incorrect. Same length != same content? 

Compare `set` of fieldnames

ankush replied:
Edge case: 1 field is added and 1 field is removed.

sagarvora replied:
`self._fields` is just a dict version of `self.fields`. In line 459 (just above this), I am filtering `field_order` based on `self._fields`. So it cannot contain anything other than the contents of `self.fields` / `self._fields`. So length match should be adequate.

ankush commented on frappe/model/meta.py:
```diff
@@ -452,14 +448,56 @@ def init_field_caches(self):
+				self._update_fields_based_on_order(field_order)
+				return
+
+			# if the first few standard fields are not in the field order, prepare to prepend them
+			if self.fields[0].fieldname not in field_order:
+				fields_to_prepend = []
+				standard_field_found = False
```
Is this because new fields on top can't have `insert_after` "guessed"?


(add in comments, if that's the case)

sagarvora replied:
Yes, that is the reason.

ankush commented on frappe/model/meta.py:
```diff
@@ -452,14 +448,56 @@ def init_field_caches(self):
+
+				for fieldname, field in self._fields.items():
+					if getattr(field, "is_custom_field", False):
+						# all custom fields from here on
+						break
+
```
Isn't there a case when custom fields can be on top, what if new standard fields are also added on top?

sagarvora replied:
The case where custom field is added to the top (in case it isn't in field order) is handled in unchanged line 507: https://github.com/frappe/frappe/pull/19822/files#diff-f23ec45599e5f8df9263a49b71658cd63d3442008ffa46e3f8aeb1b97d571556R507

This algorithm isn't meant to perfect, just _good enough_. It is possible that the custom field meant to be on the top is in the field order and the new standard fields added to the doctype will go above that. I don't think this algorithm can be perfected in the current design, which is why I had added an optional implementation of notifying users when their field order might need updates.

ankush commented:
@dj12djdjs @sagarvora this feels okay for first cut. Whatever the caveats maybe, all those exist for current code too.

Few things we can add maybe is:

- [x] Give a "Reset Layout" button to just delete the property setter (and nothing else), This will basically go back to the previous behaviour. https://github.com/frappe/frappe/pull/21290

sagarvora commented:
support for form builder added via https://github.com/frappe/frappe/pull/21297
